File: apps/minecraft/spigot/src/main/java/crabcraft/net/crabUtilities/model/MergedModelProtectionListener.java
package crabcraft.net.crabUtilities.model;

import crabcraft.net.crabUtilities.CrabMessages;
import org.bukkit.block.Crafter;
import org.bukkit.event.EventHandler;
import org.bukkit.event.EventPriority;
import org.bukkit.event.Listener;
import org.bukkit.event.block.BlockCookEvent;
import org.bukkit.event.block.CrafterCraftEvent;
import org.bukkit.event.inventory.InventoryClickEvent;
import org.bukkit.event.inventory.InventoryType;
import org.bukkit.event.inventory.PrepareAnvilEvent;
import org.bukkit.event.inventory.PrepareGrindstoneEvent;
import org.bukkit.event.inventory.PrepareItemCraftEvent;
import org.bukkit.event.inventory.PrepareSmithingEvent;
import org.bukkit.inventory.AnvilInventory;
import org.bukkit.inventory.CraftingInventory;
import org.bukkit.inventory.GrindstoneInventory;
import org.bukkit.inventory.Inventory;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.SmithingInventory;

/** Prevents recipes from losing or combining the cosmetic stored inside a merged item. */
final class MergedModelProtectionListener implements Listener {

    private final MergedModelCodec codec;

    MergedModelProtectionListener(MergedModelCodec codec) {
        this.codec = codec;
    }

    @EventHandler(priority = EventPriority.HIGHEST)
    public void onPrepareAnvil(PrepareAnvilEvent event) {
        var inventory = event.getInventory();
        if (shouldClearResult(
                event.getResult(),
                inventory.getFirstItem(),
                inventory.getSecondItem())) {
            event.setResult(null);
        }
    }

    @EventHandler(priority = EventPriority.HIGHEST)
    public void onPrepareGrindstone(PrepareGrindstoneEvent event) {
        var inventory = event.getInventory();
        if (shouldClearResult(
                event.getResult(),
                inventory.getUpperItem(),
                inventory.getLowerItem())) {
            event.setResult(null);
        }
    }

    @EventHandler(priority = EventPriority.HIGHEST)
    public void onPrepareSmithing(PrepareSmithingEvent event) {
        var inventory = event.getInventory();
        if (shouldClearResult(
                event.getResult(),
                inventory.getInputTemplate(),
                inventory.getInputEquipment(),
                inventory.getInputMineral())) {
            event.setResult(null);
        }
    }

    @EventHandler(priority = EventPriority.HIGHEST)
    public void onPrepareCrafting(PrepareItemCraftEvent event) {
        if (shouldClearResult(event.getInventory().getResult(), event.getInventory().getMatrix())) {
            event.getInventory().setResult(null);
        }
    }

    @EventHandler(priority = EventPriority.HIGHEST, ignoreCancelled = true)
    public void onBlockCook(BlockCookEvent event) {
        if (codec.isMerged(event.getSource())
                && !codec.preservesMerge(event.getSource(), event.getResult())) {
            event.setCancelled(true);
        }
    }

    @EventHandler(priority = EventPriority.HIGHEST, ignoreCancelled = true)
    public void onCrafterCraft(CrafterCraftEvent event) {
        if (event.getBlock().getState() instanceof Crafter crafter
                && shouldClearResult(event.getResult(), crafter.getInventory().getContents())) {
            event.setCancelled(true);
        }
    }

    @EventHandler(priority = EventPriority.HIGHEST, ignoreCancelled = true)
    public void onResultClick(InventoryClickEvent event) {
        if (event.getSlotType() != InventoryType.SlotType.RESULT
                || event.getClickedInventory() != event.getView().getTopInventory()
                || !hasUnsafeResult(event.getView().getTopInventory())) {
            return;
        }
        event.setCancelled(true);
        event.getWhoClicked().sendMessage(CrabMessages.error("Split the merged helmet first."));
    }

    private boolean hasUnsafeResult(Inventory inventory) {
        if (inventory instanceof AnvilInventory anvil) {
            return shouldClearResult(
                    anvil.getResult(), anvil.getFirstItem(), anvil.getSecondItem());
        }
        if (inventory instanceof GrindstoneInventory grindstone) {
            return shouldClearResult(
                    grindstone.getResult(), grindstone.getUpperItem(), grindstone.getLowerItem());
        }
        if (inventory instanceof SmithingInventory smithing) {
            return shouldClearResult(
                    smithing.getResult(),
                    smithing.getInputTemplate(),
                    smithing.getInputEquipment(),
                    smithing.getInputMineral());
        }
        if (inventory instanceof CraftingInventory crafting) {
            return shouldClearResult(crafting.getResult(), crafting.getMatrix());
        }
        return false;
    }

    private boolean shouldClearResult(ItemStack result, ItemStack... inputs) {
        ItemStack mergedInput = null;
        for (ItemStack input : inputs) {
            if (!codec.isMerged(input)) {
                continue;
            }
            if (mergedInput != null) {
                return true;
            }
            mergedInput = input;
        }

        return mergedInput != null && !codec.preservesMerge(mergedInput, result);
    }
}

File: apps/minecraft/spigot/src/main/kotlin/crabcraft/net/crabUtilities/model/MergedModelProtectionListener.kt
package crabcraft.net.crabUtilities.model

import crabcraft.net.crabUtilities.CrabMessages
import org.bukkit.block.Crafter
import org.bukkit.event.EventHandler
import org.bukkit.event.EventPriority
import org.bukkit.event.Listener
import org.bukkit.event.block.BlockCookEvent
import org.bukkit.event.block.CrafterCraftEvent
import org.bukkit.event.inventory.InventoryClickEvent
import org.bukkit.event.inventory.InventoryType
import org.bukkit.event.inventory.PrepareAnvilEvent
import org.bukkit.event.inventory.PrepareGrindstoneEvent
import org.bukkit.event.inventory.PrepareItemCraftEvent
import org.bukkit.event.inventory.PrepareSmithingEvent
import org.bukkit.inventory.AnvilInventory
import org.bukkit.inventory.CraftingInventory
import org.bukkit.inventory.GrindstoneInventory
import org.bukkit.inventory.Inventory
import org.bukkit.inventory.ItemStack
import org.bukkit.inventory.SmithingInventory

/** Prevents recipes from losing or combining the cosmetic stored inside a merged item. */
class MergedModelProtectionListener(private val codec: MergedModelCodec) : Listener {
    @EventHandler(priority = EventPriority.HIGHEST)
    fun onPrepareAnvil(event: PrepareAnvilEvent) {
        val inventory = event.inventory
        if (shouldClearResult(event.result, inventory.firstItem, inventory.secondItem)) event.result = null
    }

    @EventHandler(priority = EventPriority.HIGHEST)
    fun onPrepareGrindstone(event: PrepareGrindstoneEvent) {
        val inventory = event.inventory
        if (shouldClearResult(event.result, inventory.upperItem, inventory.lowerItem)) event.result = null
    }

    @EventHandler(priority = EventPriority.HIGHEST)
    fun onPrepareSmithing(event: PrepareSmithingEvent) {
        val inventory = event.inventory
        if (shouldClearResult(event.result, inventory.inputTemplate, inventory.inputEquipment, inventory.inputMineral)) event.result = null
    }

    @EventHandler(priority = EventPriority.HIGHEST)
    fun onPrepareCrafting(event: PrepareItemCraftEvent) {
        if (shouldClearResult(event.inventory.result, *event.inventory.matrix)) event.inventory.result = null
    }

    @EventHandler(priority = EventPriority.HIGHEST, ignoreCancelled = true)
    fun onBlockCook(event: BlockCookEvent) {
        if (codec.isMerged(event.source) && !codec.preservesMerge(event.source, event.result)) event.isCancelled = true
    }

    @EventHandler(priority = EventPriority.HIGHEST, ignoreCancelled = true)
    fun onCrafterCraft(event: CrafterCraftEvent) {
        val crafter = event.block.state as? Crafter
        if (crafter != null && shouldClearResult(event.result, *crafter.inventory.contents)) event.isCancelled = true
    }

    @EventHandler(priority = EventPriority.HIGHEST, ignoreCancelled = true)
    fun onResultClick(event: InventoryClickEvent) {
        if (event.slotType != InventoryType.SlotType.RESULT
            || event.clickedInventory != event.view.topInventory || !hasUnsafeResult(event.view.topInventory)) return
        event.isCancelled = true
        event.whoClicked.sendMessage(CrabMessages.error("Split the merged helmet first."))
    }

    private fun hasUnsafeResult(inventory: Inventory): Boolean = when (inventory) {
        is AnvilInventory -> shouldClearResult(inventory.result, inventory.firstItem, inventory.secondItem)
        is GrindstoneInventory -> shouldClearResult(inventory.result, inventory.upperItem, inventory.lowerItem)
        is SmithingInventory -> shouldClearResult(inventory.result, inventory.inputTemplate, inventory.inputEquipment, inventory.inputMineral)
        is CraftingInventory -> shouldClearResult(inventory.result, *inventory.matrix)
        else -> false
    }

    private fun shouldClearResult(result: ItemStack?, vararg inputs: ItemStack?): Boolean {
        var mergedInput: ItemStack? = null
        for (input in inputs) {
            if (!codec.isMerged(input)) continue
            if (mergedInput != null) return true
            mergedInput = input
        }
        return mergedInput != null && !codec.preservesMerge(mergedInput, result)
    }
}
